File: 114.flatten-binary-tree-to-linked-list.cpp
/*
 * @lc app=leetcode id=114 lang=cpp
 *
 * [114] Flatten Binary Tree to Linked List
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void get_element(TreeNode *root, vector<TreeNode*>& res){
        if (root){
            res.push_back(root);
            get_element(root->left, res);
            get_element(root->right, res);
        }
    }
    void flatten(TreeNode* root) {
        if(!root){
            return;
        }
        vector<TreeNode*> res;
        get_element(root, res);
        TreeNode* fake_head = new TreeNode(1);
        TreeNode* cur = fake_head;
        for(int i = 0; i < res.size(); ++i){
            cur->left = nullptr;
            cur->right = res[i];
            cur = cur->right;
        }
        cur->left = nullptr;
        cur->right = nullptr;
    }
};
// @lc code=end

File: 114.flatten-binary-tree-to-linked-list.py
#
# @lc app=leetcode id=114 lang=python3
#
# [114] Flatten Binary Tree to Linked List
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right



class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if root is None:
            return 
        tmp_root = TreeNode(1)
        cur_node = tmp_root

        def flatten_helper(root_):
            nonlocal cur_node
            if root_ is None:
                return
            cur_node.right = TreeNode(root_.val)
            cur_node = cur_node.right
            flatten_helper(root_.left)
            flatten_helper(root_.right)
            return
        flatten_helper(root)
        root.left = None
        root.right = tmp_root.right.right
        return
        
# @lc code=end

